go/vt/vttablet/tabletserver/vstreamer/tablestreamer.go

```diff
@@ -0,0 +1,176 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vstreamer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"vitess.io/vitess/go/sqlescape"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/dbconfigs"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/schema"
+
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+)
+
+/*
+	TableStreamer is a VStreamer that streams all tables in a keyspace. It iterates through all tables in a keyspace
+	and streams them one by one. It is not resilient: if there is any error that breaks the stream, for example,
+	reparenting or a network error, it will not recover and a new workflow will have to be created.
+*/
+
+// TableStreamer exposes an externally usable interface to tableStreamer.
+type TableStreamer interface {
+	Stream() error
+	Cancel()
+}
+
+type tableStreamer struct {
+	ctx    context.Context
+	cancel func()
+
+	cp      dbconfigs.Connector
+	se      *schema.Engine
+	send    func(*binlogdatapb.VStreamTablesResponse) error
+	vschema *localVSchema
+	vse     *Engine
+
+	snapshotConn *snapshotConn
+	tables       []string
+	gtid         string
+}
+
+func newTableStreamer(ctx context.Context, cp dbconfigs.Connector, se *schema.Engine, vschema *localVSchema, send func(response *binlogdatapb.VStreamTablesResponse) error, vse *Engine) *tableStreamer {
+	ctx, cancel := context.WithCancel(ctx)
+	return &tableStreamer{
+		ctx:     ctx,
+		cancel:  cancel,
+		cp:      cp,
+		se:      se,
+		send:    send,
+		vschema: vschema,
+		vse:     vse,
+	}
+}
+
+func (ts *tableStreamer) Cancel() {
+	log.Info("TableStreamer Cancel() called")
+	ts.cancel()
+}
+
+func (ts *tableStreamer) Stream() error {
+	// Ensure that the schema engine is Open. If vttablet came up in a non_serving role, it may not have been initialized.
+	var err error
+	if err = ts.se.Open(); err != nil {
+		return err
+	}
+
+	conn, err := snapshotConnect(ts.ctx, ts.cp)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	ts.snapshotConn = conn
+
+	if ts.gtid, err = conn.startSnapshotAllTables(ts.ctx); err != nil {
+		return err
+	}
+
+	if _, err := conn.ExecuteFetch("set names 'binary'", 1, false); err != nil {
+		return err
+	}
+
+	rs, err := conn.ExecuteFetch("show tables", -1, true)
+	if err != nil {
+		return err
+	}
+	for _, row := range rs.Rows {
+		tableName := row[0].ToString()
+		ts.tables = append(ts.tables, tableName)
+	}
+	log.Infof("Found %d tables to stream: %+v", len(ts.tables), ts.tables)
```

Comment:
`ts.tables` is a string slice so strings.Join(ts.tables, ",") would be better IMO (and we don't need +v as there are no members).